### DISCUSSION TITLE
Bug: Wrong type for onSelect arguments when passing typed mode property

### DISCUSSION THREAD

**Discussion by VitaliyR** [gpbl/react-day-picker#1570] [Support] [answered]
### Bug description

Bug with types for the react-day-picker

### To reproduce

https://codesandbox.io/s/react-daypicker-forked-wdgtko?file=/src/App.tsx

#### Steps

Write the code

### Expected behavior

For this case argument `date` should have type `Date` which comes from `SelectSingleEventHandler`

### Screenshots

<img width="1278" alt="Untitled" src="https://user-images.githubusercontent.com/2149612/193266014-bffe2e6a-1050-40f7-a5a7-16b7d1d3e415.png">

### Additional context

I assume this happens as `DayPickerDefaultProps` has dynamic `mode` usage which breaks union, but not exactly sure.

**Comment by gpbl** [comment]
Hey thanks for reporting this, I see some issues as well but I'm not sure if they are related to the discriminated union. Will investigate...

**Comment by gpbl** [comment]
This should fix it: https://github.com/gpbl/react-day-picker/pull/1571

**Reply by VitaliyR** [reply]
@gpbl sadly 8.3.1 not helps in the provided codesandbox - see the link above, still erroring with same `any` error, so types are not picked up properly

**Reply by gpbl** [reply]
@VitaliyR thanks for the feedback! Uhm it does work here - could you wait the Sandbox to download the dependencies?

<img width="1153" alt="Screenshot 2022-10-03 at 06 29 08" src="https://user-images.githubusercontent.com/120693/193566711-8931ea7e-31bd-49c5-b7b2-14abd750cfec.png">

**Reply by VitaliyR** [reply]
Indeed. I've updated an attributes to more real life example when they are coming from the props.
Weirdly mode isn't picked up if it comes from the `Props`. Perhaps you know why is that happening?

**Reply by gpbl** [reply]
@VitaliyR there, the typescript compiler cannot infer the value of `props.mode`, thus the type error. You want to write some conditionals, to help the compiler understand what is `props.mode` and get the correct type:

```tsx
if (props.mode === 'single')
 <DayPicker mode={props.mode} onSelect={...} />
```

**Reply by VitaliyR** [reply]
Yeah, could not fight anylonger with TS and went with `if` battery :) 
But original fix still fixed some things, so thanks a lot for it!

**Comment by bartlomiej-bykowy-jt** [comment]
Hi! I'm not opening new issue because I think it can be somehow related with this one 🙂 

So, after updating to 8.3.1 this issue appeared:
![Screenshot 2022-10-05 at 12 00 00](https://user-images.githubusercontent.com/106966230/194036711-c4b3b6fb-c0bc-4b0c-b3d9-951cf83767bf.png)

`mode` and `selectedDate` comes from props. Their types are defined as follows:
```
mode?: 'single' | 'multiple' | 'range';
selectedDate: DatesType
```
(and `DatesType: Date | Date[] | DateRange | undefined`, where `DateRange` is imported from `react-day-picker)

At this point I ran out of ideas how to resolve this issue. The only solution that worked, was something like this:
```
{ mode === 'single' && <DayPicker mode='single' selected={selectedDate as Date | undefined} {...otherProps} /> }
{ mode === 'multiple' && <DayPicker mode='multiple' selected={selectedDate as Date[] | undefined} {...otherProps} /> }
{ mode === 'range' && <DayPicker mode='range' selected={selectedDate as DateRange | undefined} {...otherProps} /> }
```
but to be hones it doesn't look too clean.

Maybe you would know what is the issue here and if there is a better way to resolve it than the one I showed above?

Thanks in advance 😃

**Reply by gpbl** [reply]
Thanks for the report, definitely it shouldn't work this way! Will investigate...

**Reply by owaisahmed5300** [reply]
Any news?

**Comment by gpbl** [comment, marked as answer]
To close this discussion, you should use type guards, e.g.

```tsx
import { DayPicker, type PropsMulti, type PropsSingle } from "react-day-picker";
import "react-day-picker/dist/style.css";

type AppProps = PropsSingle | PropsMulti;

// Type guard to check if props is for single mode
function isSingleMode(props: AppProps): props is PropsSingle {
  return props.mode === "single";
}

export default function App(props: AppProps) {
  return isSingleMode(props) ? (
    <DayPicker
      mode="single"
      selected={props.selected}
      showWeekNumber
      onSelect={props.onSelect}
    />
  ) : (
    <DayPicker
      mode="multiple"
      selected={props.selected}
      showWeekNumber
      onSelect={props.onSelect}
    />
  );
}
```

**Reply by andreasvirkus** [reply]
What if in both scenarios there's also an optional `required` prop, should that result in a 4-branch rendering condition now? This doesn't seem like optimal DX or maybe I'm missing something... 🤔

**Reply by gpbl** [reply]
@andreasvirkus Yeah the DX there is not that nice. It is really challenging to get these types right. They are defined in [props.ts](https://github.com/gpbl/react-day-picker/blob/main/src/types/props.ts#L28-L37).

Feel free to suggest any improvements if you have better ideas - I've tried many alternatives without luck - this is the best solution I could find.
